File: csharpcore60/ShiMart/ShiMart.cs

using System.Collections.Generic;

namespace ShiMartKata
{
    public class ShiMart
    {
        IList<Item> Items;
        public ShiMart(IList<Item> Items)
        {
            this.Items = Items;
        }

        // TODO:Increase code readbility
        // 1. Reduce if conditions
        // 2. Refacor code in more methods
        // 3. Refactor SellIn logic in a new method : UpdateSellIn. Note: QualityCalulation is dependent on SellIn Value


        // Make the program item independent 
        // 1. Item.cs in a class ItemWithType.cs, inherits Item.cs
        // 2. Use Builder pattern for BuildingItem
        // 3. It requires creation of centeraliztion of multiple type of items and handle related busisness logic,
        // I would prefer, use Abstarct Factory Pattern for different type of items        
        public void UpdateQuality()
        {
            for (var i = 0; i < Items.Count; i++)
            {
                if (Items[i].Name != "Canned Beans")
                {
                    Items[i].SellIn = Items[i].SellIn - 1;
                }

                if (Items[i].Name == "Aged Brie")
                {
                    if (Items[i].Quality < 50)
                    {
                        Items[i].Quality = Items[i].Quality + 1;
                    }
                }
                else
                {
                    if (Items[i].Quality > 0)
                    {
                        if (Items[i].Name != "Canned Beans" && !Items[i].Name.Contains("Baked"))
                        {
                            Items[i].Quality = Items[i].Quality - 1;
                        }

                        if (Items[i].Name.Contains("Baked"))
                        {
                            Items[i].Quality = Items[i].Quality - 2;
                        }
                    }
                }

                if (Items[i].SellIn < 0)
                {
                    if (Items[i].Name == "Aged Brie")
                    {
                        if (Items[i].Quality < 50)
                        {
                            Items[i].Quality = Items[i].Quality + 1;
                        }
                    }
                    else
                    {
                        if (Items[i].Quality > 0)
                        {
                            if (Items[i].Name != "Canned Beans" && !Items[i].Name.Contains("Baked"))
                            {
                                Items[i].Quality = Items[i].Quality - 1;
                            }
                            if (Items[i].Name.Contains("Baked"))
                            {
                                Items[i].Quality = Items[i].Quality - 2;
                            }
                        }
                    }
                }
            }
        }
    }
}

File: TypeScript/app/shi-mart.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

export class ShiMart {
  items: Array<Item>;

  constructor(items: Array<Item> = []) {
    this.items = items;
  }

  updateQuality() {
    for (let i = 0; i < this.items.length; i++) 
    {
      if (this.items[i].name !== "Canned Beans") 
      {
        this.items[i].sellIn = this.items[i].sellIn - 1;
      }

      if (this.items[i].name === "Aged Brie")
      {
        if (this.items[i].quality < 50)
        {
          this.items[i].quality = this.items[i].quality + 1;
        }
      } else {
        if (this.items[i].quality > 0) 
        {
          if (this.items[i].name !== "Canned Beans") 
          {
            this.items[i].quality = this.items[i].quality - 1;
          }
        }
      }

      if (this.items[i].sellIn < 0) 
      {
        if (this.items[i].name === "Aged Brie") 
        {
          if (this.items[i].quality < 50) 
          {
            this.items[i].quality = this.items[i].quality + 1;
          }
        } else {
          if (this.items[i].quality > 0)
          {
            if (this.items[i].name !== "Canned Beans")
            {
              this.items[i].quality = this.items[i].quality - 1;
            }
          }
        }
      }
    }
  }
}